### PR TITLE
new(proposals): driver kernel testing framework

### PR DESCRIPTION
**Pull request by incertum**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

/kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

> /area libsinsp

> /area tests

/area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

Formalize a driver kernel testing framework.
@falcosecurity/libs-maintainers 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: proposals/20230530-driver-kernel-testing-framework.md
Comment (incertum) [reply proposals/20230530-driver-kernel-testing-framework.md:135]: Fair point, can tweak.

Perhaps another good discussion point: `s390x ` is supported by libs but not officially by Falco. Maybe I can say that because of this initially we prioritize x86 and aarch64 tests, subject to change in future iterations of the test framework ...

@hbrueckner would you have ideas on how we can test `s390x ` specifically for libs better? And ideally cut down the durations of s390x libs builds as well, they take a fairly long time atm. Thank you!

Comment (leogr) [reply proposals/20230530-driver-kernel-testing-framework.md]: My recommendation is:

- Falco officially supported architectures should be prioritized by their adoption level (ie. `x86_64` is P0, `aarch64` is P1)
- just best effort (low priority) for any other arch supported by libs (ie. `s390x` is Nice-To-Have).

Comment (incertum) [reply proposals/20230530-driver-kernel-testing-framework.md]: Love it, this way we clearly articulate priorities.

File: proposals/20230530-driver-kernel-testing-framework.md
Comment (maxgio92) [inline proposals/20230530-driver-kernel-testing-framework.md:218]: Just as a possible future improvement, we could think about having the kernels grid dynamic.

Comment (incertum) [reply proposals/20230530-driver-kernel-testing-framework.md:218]: ACK, I'll work this in with the next cleanup commit as we are collecting feedback.

File: proposals/20230530-driver-kernel-testing-framework.md
Comment (maxgio92) [inline proposals/20230530-driver-kernel-testing-framework.md:211]: We could consider to spin up VM with KVM on hosted GitHub runners, or with something like [actuated](https://github.com/orgs/community/discussions/8305#discussioncomment-5011953) (on microVMs).

**Comment by maxgio92** [comment]
Love this proposal, @incertum, thank you for this great work!

**Comment by TheFoxAtWork** [comment]
Love this!  Thank you so much for pulling this together!

**Comment by incertum** [comment]
> Love this! Thank you so much for pulling this together!

Thank you @TheFoxAtWork!

File: proposals/20230530-driver-kernel-testing-framework.md
Comment (incertum) [reply proposals/20230530-driver-kernel-testing-framework.md:218]: @maxgio92 I have worked your suggestions in.

**Comment by incertum** [comment]
@falcosecurity/libs-maintainers I have moved this PR out of WIP, it is now ready for final review. Thanks in advance!

File: proposals/20230530-driver-kernel-testing-framework.md
Comment (maxgio92) [reply proposals/20230530-driver-kernel-testing-framework.md:218]: Thank you :)

File: proposals/20230530-driver-kernel-testing-framework.md
Comment (FedeDP) [inline proposals/20230530-driver-kernel-testing-framework.md:30]: i'd use `kernel releases` here; that's what we use elsewhere, and what test-infra dbg, driverkit and kernel-crawler use.

File: proposals/20230530-driver-kernel-testing-framework.md
Comment (FedeDP) [inline proposals/20230530-driver-kernel-testing-framework.md:94]: We also nightly-test the build against latest mainline kernel now: https://github.com/falcosecurity/libs/blob/master/.github/workflows/latest-kernel.yml

File: proposals/20230530-driver-kernel-testing-framework.md
Comment (FedeDP) [inline proposals/20230530-driver-kernel-testing-framework.md:100]: All of these are on x86_64. 
I am looking into adding same tests for arm64 too.

**Review by FedeDP — COMMENTED** [review]
Left some comments; LGTM, thank you @incertum ! Great work!

A minor question: are we talking about adding nightly tests? I think we do not want to run these tests on every PR/master change.
Perhaps weekly tests can be enough too.
Thank you again btw!

File: proposals/20230530-driver-kernel-testing-framework.md
Comment (incertum) [reply proposals/20230530-driver-kernel-testing-framework.md]: :) adjusted and slick btw!

**Comment by incertum** [comment]
> A minor question: are we talking about adding nightly tests? I think we do not want to run these tests on every PR/master change. Perhaps weekly tests can be enough too. Thank you again btw!

where I mentioned it to be "on demand" initially I added "(such as nightly tests)"

**Comment by incertum** [comment]
What needs to happen for us to move forward with this?

**Comment by incertum** [comment]
> What needs to happen for us to move forward with this?

@TheFoxAtWork as a first step as suggested we created a new dedicated CNCF Service Desk ticket https://cncfservicedesk.atlassian.net/servicedesk/customer/portal/1/CNCFSD-1822. Thank you!

File: proposals/20230530-driver-kernel-testing-framework.md
Comment (leogr) [inline proposals/20230530-driver-kernel-testing-framework.md:230]:
```suggestion
```

We want to abandon [CircleCI in favor of GitHub Action](https://github.com/falcosecurity/falco/issues/2454)

**Review by leogr — COMMENTED** [review]
@incertum Thank you so much for this amazing proposal. :pray: 

LGTM, just left a minor suggestion.

I'll give it a second look early next week then I will give my final approval.

**Review by leogr — APPROVED** [review]
LGTM :+1:

File: proposals/20230530-driver-kernel-testing-framework.md
Comment (Andreagit97) [inline proposals/20230530-driver-kernel-testing-framework.md:77]: > run is positive and that the drivers_test unit tests

Just one minor note about this. We know that on older kernel versions `~<5.0` driver_tests could fail due to conflicting events captured. Let me explain better, is possible that while we search for a `close` event spawned by our test we face another `clone` event generated by the system causing a test failure (as far as i saw this is very frequent when we use `tracepoints` instead of `raw_traceoints` in the old probe, so in kernels between `4.14` and `4.17`) ... so probably we won't be able to rely on the drivers_test on all machines as a source of truth. We will try to fix driver_tests to avoid these failures but in the meanwhile, I would consider scap-open as the only source of truth and use the drivers_tests as a best-effort check. WDYT?

**Review by Andreagit97 — APPROVED** [review]
/approve

I left just one comment on the `driver_tests` part. Please unhold if you are fine with this :)
Thank you for the amazing job!

/hold

File: proposals/20230530-driver-kernel-testing-framework.md
Comment (incertum) [reply proposals/20230530-driver-kernel-testing-framework.md:77]: Thanks @Andreagit97 I agree, `driver_tests` should be best effort at first and we can stabilize it subsequently.

**Pull request merged**
